### DISCUSSION TITLE
AWS Amplify & Vite: global is not defined

### DISCUSSION THREAD

**Discussion by BabyDino** [nuxt/nuxt#16009] [Questions] [answered]
Hi all,

I am trying to get AWS Amplify to work in Nuxt 3. I got vee-validate v4 working by doing some imports in a plugin. Then I tried to do the same trick with the aws-amplify library. That's where I ran into issues. This is my plugin code:

```
import { defineNuxtPlugin } from '#app'
import { useRuntimeConfig } from '#app'
import { Amplify } from '@aws-amplify/core'
import { Analytics } from '@aws-amplify/analytics'
import { Auth } from '@aws-amplify/auth'
import { API } from '@aws-amplify/api'
import { Storage } from '@aws-amplify/storage'

export default defineNuxtPlugin(() => {
    const config = useRuntimeConfig();

    const amplifyConfig = {
        Auth: {
            region: config.APP_REGION,
            ...
        },
        Storage: {
            AWSS3: {
            bucket: config.S3_BUCKET,
            region: config.APP_REGION,
            },
        },
        ...
    }

    Amplify.configure(amplifyConfig);
  })
```

An error occurs:

```
Uncaught ReferenceError: global is not defined
```

This is the relevant code in the Amplify library/AWS SDK:

```
Buffer.TYPED_ARRAY_SUPPORT = global.TYPED_ARRAY_SUPPORT !== undefined
  ? global.TYPED_ARRAY_SUPPORT
  : typedArraySupport()
```
Apparently this can be solved by adding `(window as any).global = window;` somewhere (haven't found out where yet).

Can anyone confirm this is related to #1922? If so, does anyone have a temporary workaround for this?

Thanks in advance!

**Comment by mt-u** [comment, marked as answer]
Hi.
[This comment](https://github.com/nuxt/framework/issues/2070#issue-1059957544) and [his repository(fix branch)](https://github.com/sweepy84/nuxt3/tree/fix) may be useful to you.
I was struggling as well, but finally I have been able to get the Amplify UI to work with Nuxt3 in his way.
However, when I used it as it was, the meta tag setting caused a 404 error, so I moved the `polyfills.js` to the Nuxt3's `public` folder and changed the reference path of script's src in the `nuxt.config.ts` as well.

@sweepy84, thank you for your great job!

**Reply by BabyDino** [reply]
I totally missed that issue while searching for amplify. Many thanks for guiding me. I will def test this. Thanks again!

~~Do you know of an amplify-js issue related to this as well?~~ nvm, found issues: 

- https://github.com/aws-amplify/amplify-ui/issues/268#issuecomment-953375909
- https://github.com/aws/aws-sdk-js/issues/3673

**Reply by BabyDino** [reply]
It seems to work when you do a configure and only a configure. It looks that the plugins are loaded first, before the meta.

If you return anything, or do anything else with the imports the global issue comes back again. I have to figure out what the exact lifecycle is. In our case, with a custom API, this won't work:

```
import { defineNuxtPlugin } from '#app';
import { useRuntimeConfig } from '#app';
import { Amplify } from '@aws-amplify/core';
import { Auth } from '@aws-amplify/auth';

export default defineNuxtPlugin(() => {
  const config = useRuntimeConfig();

  const amplifyConfig = {
    Auth: {
      region: config.APP_REGION,
      identityPoolRegion: config.APP_REGION,
      identityPoolId: config.USER_IDENTITY_POOL_ID,
      userPoolId: config.USER_POOL_ID,
      userPoolWebClientId: config.USER_POOL_CLIENT_ID
    },
    API: {
      graphql_endpoint: config.API_URL + '/graphql',
      graphql_headers: async () => {
        const session = await Auth.currentSession();
        const token = session.getIdToken().getJwtToken();

        return {
          Authorization: token
        };
      }
    }
  };

  console.log(amplifyConfig);

  Amplify.configure(amplifyConfig);

  return {
    auth: Auth
    // storage: Storage,
    // api: API,
    // analytics: Analytics
  };
});
```

**Comment by mt-u** [comment]
Following [the Nuxt3 guide](https://v3.nuxtjs.org/docs/directory-structure/plugins#automatically-providing-helpers), I succeeded in providing the functionality of the plugin to the component as follows:
```typescript
// plugins/amplify.ts
import { defineNuxtPlugin } from '#app';

import awsConfig from '../aws-exports';
import { Amplify } from '@aws-amplify/core';
import { Auth } from '@aws-amplify/auth';

export default defineNuxtPlugin(nuxtApp => {
  Amplify.configure(awsConfig);

  return {
    provide: {
      auth: Auth,
    },
  };
});
```
I could use it in the component as follows:
```vue
<!-- app.vue -->
<template>
  <div>
    <Script src="./js/polyfills.js"/>
    <Link rel="preconnect" href="https://fonts.googleapis.com"/>
    <Link rel="preconnect" href="https://fonts.gstatic.com" crossorigin/>
    <Link
        href="https://fonts.googleapis.com/css2?family=Inter:slnt,wght@-10..0,100..900&display=swap"
        rel="stylesheet"
    />

    <Authenticator>
      <template v-slot="{ user, signOut }">
        <h1>Hello {{ user.username }}!</h1>
        <button @click="signOut">Sign Out</button>
        <NuxtPage/>
      </template>
    </Authenticator>

    <p>{{ currentAuthenticatedUser ? currentAuthenticatedUser .username : 'You are unauthorized.' }}</p>
  </div>
</template>

<script setup lang="ts">
import { Authenticator } from "@aws-amplify/ui-vue";
import "@aws-amplify/ui-vue/styles.css";

// Here, the auth object from the plugin is injected and used.
const { $auth } = useNuxtApp();
let currentAuthenticatedUser = undefined;
try {
  currentAuthenticatedUser = await $auth.currentAuthenticatedUser();
} catch (err) {
  console.log(err);
}
</script>
```
I hope this helps you.

**Reply by BabyDino** [reply]
Thanks again. I think I am on to something. It seems that `console.log` in the plugin is always executed earlier than the script load in templates/components.

I am using pages instead of app.vue (I was also using `meta` in the config). Maybe the issue lies there, because my code is exactly the same as yours. The minute I `return` (with `provide`, thanks for pointing that out) in `amplify.ts` I get the `global is not defined error` again.

If I add a default layout with the `<Script>` tag it doesn't help, because the plugin load seems to be earlier than page load.

I am using `ssr: false`, by the way.

Update: `ssr: false` seems to be the issue. I have some bootstrap 5 issues and an import broke it. I will test further. Thanks again.

**Reply by mt-u** [reply]
I'm sorry it didn't help.
Oh, what is the cause? Maybe upgrading Nuxt 3 will solve it.
My environment is as follows.
```bash
$ node --version
v14.18.1

$ npm list --depth=0
my-nuxt-amplify-project
+-- @aws-amplify/ui-vue@2.0.5
+-- aws-amplify@4.3.10
`-- nuxt3@3.0.0-27324955.23397e6
```

> I get the `global is not defined error` again.

Is this error displayed in the browser's developer tools console?
If so, I suspect the `polyfills.js` isn't loaded.
Could you please check if the `src` attribute of the `script` tag actually output on the browser is the correct URL of `polyfills.js`?
Also, please check if you can actually access the URL with a browser and display it?
In my case, the location where `polyfills.js` is located is `public/js/polyfills.js`([About `public` directory](https://v3.nuxtjs.org/docs/directory-structure/public)), so it should be output as follows:
```html
<!-- on browser -->
<script src="./js/polyfills.js"></script>
```
Otherwise I'll get a 404 error.

By the way, I also use the `pages` directory and the `NuxtPage` tag.
Also, even if I rewrite it to use the `meta` config as follows instead of the `Script` tag of Nuxt3, the error does not occur.
In addition, I tried calling `console.log` in the `plugins/amplify.ts` and it didn't cause any problems.
```typescript
// nuxt.config.ts
export default defineNuxtConfig({
  // other settings...

  meta: {
    script: [
      { src: './js/polyfills.js' },
    ],
  },

  // other settings...
})
```

**Reply by BabyDino** [reply]
I've got it working. My issue was non Amplify related, your fix works. 

Short version: I had a bootstrap 5 import which broke everything. With `ssr: off` this was fixed. But by disabling SSR, Amplify keep throwing me the global issue, which has something to do with this issue: #1922.

I now load bootstrap 5 via meta scripts, and this works fine for now. I do not get any more errors, I will now focus on Amplify. I am a bit surprised it works with SSR.

**Reply by ennioVisco** [reply]
> Following [the Nuxt3 guide](https://v3.nuxtjs.org/docs/directory-structure/plugins#automatically-providing-helpers), I succeeded in providing the functionality of the plugin to the component as follows:
> 
> ```ts
> // plugins/amplify.ts
> import { defineNuxtPlugin } from '#app';
> 
> import awsConfig from '../aws-exports';
> import { Amplify } from '@aws-amplify/core';
> import { Auth } from '@aws-amplify/auth';
> 
> export default defineNuxtPlugin(nuxtApp => {
>   Amplify.configure(awsConfig);
> 
>   return {
>     provide: {
>       auth: Auth,
>     },
>   };
> });
> ```
> 
> I could use it in the component as follows: [...]

It's weird because @mt-u 's fix makes perfect sense, but it seems that Amplify is not really ready for it, as it gives me the following error:
```
node_modules\amazon-cognito-identity-js\es\utils\cryptoSecureRandomInt.js:29
export default function cryptoSecureRandomInt() {
^^^^^^
SyntaxError: Unexpected token 'export'
```
I am, as well, just using the Auth module.
Is it still working to you? Are you maybe using a specific version?

**Reply by mt-u** [reply]
Hi @ennioVisco.
It's been a while, so I tried running it, remembering it little by little, and confirmed that my sample project was working properly.
The environment and package version are as written in [my comment above](https://github.com/nuxt/framework/discussions/2308#discussioncomment-1830479).

**Reply by mt-u** [reply]
Then, after updating the npm packages to the latest version, I tried again and confirmed that it was still working properly.
The version is as follows.
```
$ npm list --depth=0
my-nuxt-amplify-project
+-- @aws-amplify/ui-vue@2.1.2
+-- aws-amplify@4.3.14
`-- nuxt3@3.0.0-27418641.56aabd6
```

**Reply by ennioVisco** [reply]
> ```ts
> import { defineNuxtPlugin } from '#app';
> 
> import awsConfig from '../aws-exports';
> import { Amplify } from '@aws-amplify/core';
> import { Auth } from '@aws-amplify/auth';
> 
> export default defineNuxtPlugin(nuxtApp => {
>   Amplify.configure(awsConfig);
> 
>   return {
>     provide: {
>       auth: Auth,
>     },
>   };
> });
> ```

Unfortunately, even in your environment, I keep running in the same issue. @mt-u, you can find [here a Sandbox environment](https://codesandbox.io/s/rough-sunset-wqnvsl) that tries to reproduce your fix with minimal code. Do you see something wrong with it? It could also be beneficial to future readers of this thread 😊

**Reply by mt-u** [reply]
@ennioVisco I have seen your code.
First of all, I'm getting an error because apparently the `polyfills.js` isn't included.
I mentioned the file in [this comment above](https://github.com/nuxt/framework/discussions/2308#discussioncomment-1811050).
Please refer to [this link](https://github.com/sweepy84/nuxt3/blob/fix/polyfills.js) for the contents of `polyfills.js`.
I put this file in `[project_root]/public/js/polyfills.js` and loaded it using the `Script` tag in the `app.vue`.
Please refer to the contents of `app.vue` at the bottom of this reply.

And then, the content of `aws-exports.js` seems to be different from what is generated by `amplify init`, `amplify add auth` and `amplify push`. So, you may need to fix it according to [the official Amplify guide](https://docs.amplify.aws/lib/auth/start/q/platform/js/#re-use-existing-authentication-resource).

Finally, regarding how to use the `amplify-auth` plugin, it seems that calling a plugin function within the context of `<script setup>` will result in an error. However, it doesn't seem to be a problem if it's inside the `onMounted` lifecycle hook, so I've confirmed that it works fine as follows.
``` vue
<!-- app.vue -->
<template>
  <div>
    <Script src="./js/polyfills.js"/>
    <p>authenticated: {{ authenticated }}</p>
  </div>
</template>

<script setup lang="ts">
const { $auth } = useNuxtApp();

const authenticated = ref(false);

onMounted(async function () {
  authenticated.value = !!(await $auth.currentAuthenticatedUser());
});
</script>
```

**Reply by mt-u** [reply]
@ennioVisco Yet one more thing, I think there is a problem with your `composables/useAuth.ts`, this is probably not directly related to the error you are facing.
[`Auth.currentAuthenticatedUser()` returns a `Promise`](https://aws-amplify.github.io/amplify-js/api/classes/authclass.html#currentauthenticateduser), so it doesn't make sense to compare it to `undefined`. I think it should be changed to one that takes `Promise` into consideration, but there are two concerns.

- Since [the Nuxt3 manual](https://v3.nuxtjs.org/docs/usage/nuxt-app) states that `useNuxtApp()` works only within `setup`, plugins, or lifecycle hooks, using plugins within composables may be incorrect.
- I don't know if it is the correct usage to handle `Promise` directly by `useState`. `useState` returns `Ref<T>`, so it may be okay, but I think it's complicated to use.

Based on these, it may be simpler to handle the `bool` value in `useState` and prepare a separate update function as follows.
``` typescript
// composables/useAuth.ts
export default function () {
  const { $auth } = useNuxtApp();

  const isAuthenticated = useState('authenticated', () => false);

  return {
    isAuthenticated,
    refresh: async () => {
      isAuthenticated.value = !!(await $auth.currentAuthenticatedUser());
    },
  };
}
```

Thankyou!

**Reply by ennioVisco** [reply]
@mt-u first of all thanks a lot for your precision and kindness, hopefully this effort will be appreciated by other readers stumbling in this issue. I'll be sure to update the related issue and discussion on nuxt3 to be sure that they point in the right direction.

> @ennioVisco I have seen your code. First of all, I'm getting an error because apparently the `polyfills.js` isn't included. I mentioned the file in [this comment above](https://github.com/nuxt/framework/discussions/2308#discussioncomment-1811050). Please refer to [this link](https://github.com/sweepy84/nuxt3/blob/fix/polyfills.js) for the contents of `polyfills.js`. I put this file in `[project_root]/public/js/polyfills.js` and loaded it using the `Script` tag in the `app.vue`. Please refer to the contents of `app.vue` at the bottom of this reply.

You are absolutely right, I missed the point with polyfill.js, as the name made me think it was needed for amplify's UI package, which I wanted to exclude for this minimal example. 

> And then, the content of `aws-exports.js` seems to be different from what is generated by `amplify init`, `amplify add auth` and `amplify push`. So, you may need to fix it according to [the official Amplify guide](https://docs.amplify.aws/lib/auth/start/q/platform/js/#re-use-existing-authentication-resource).

You are absolutely right, but the reason for that is that I don't want to use Amplify on the server-side, but just Amplify's SDK for interacting with Cognito. I have this working with the very same settings. Nevertheless, even changing them to the format of the Amplify guide doesn't seem to change the issue (it is way before in the call stack I presume).

> Finally, regarding how to use the `amplify-auth` plugin, it seems that calling a plugin function within the context of `<script setup>` will result in an error. However, it doesn't seem to be a problem if it's inside the `onMounted` lifecycle hook[...].

Good point, didn't note this, good catch. I'll stick to this solution for the [sandbox](https://codesandbox.io/s/rough-sunset-wqnvsl?file=/app.vue), which looks simpler.
About the composable I wrote, your point is again very valid: I just wanted some minimal code to reproduce the issue, and that one seemed reasonable, but your correction makes much more sense.

Overall, unfortunately, I still incur in the exact same bug, which you can find in the updated sandbox environment [here](https://codesandbox.io/s/rough-sunset-wqnvsl?file=/app.vue).

**Reply by ennioVisco** [reply]
Quick heads up, it turned out I was missing one last element, which is forcing the rollup transpiling of the identity module of Amplify, this line of code makes the magic:
```ts
// nuxt.config.ts
import { defineNuxtConfig } from "nuxt3";

export default defineNuxtConfig({
  build: {
    transpile: ["amazon-cognito-identity-js"]
  }
});

```

I posted a complete summary of the solution here: https://github.com/nuxt/framework/issues/2874#issuecomment-1046285059, and pointed to it and to this thread in https://github.com/nuxt/framework/issues/2070.

**Reply by mt-u** [reply]
@ennioVisco I read your issue https://github.com/nuxt/framework/issues/2874#issuecomment-1046285059. It's a great job. thank you.
And your information made me realize that I had missed it.

First, I have to apologize to you. I wasn't aware of `nuxt.config.ts` and forgot to mention it. I'm really sorry.

My `nuxt.config.ts` is as follows.
``` typescript
// nuxt.config.ts
import { defineNuxtConfig } from 'nuxt3'

export default defineNuxtConfig({
  vite: {
    resolve: {
      alias: {
        // Required to build as per https://github.com/aws-amplify/amplify-ui/issues/268#issuecomment-953375909
        './runtimeConfig': './runtimeConfig.browser',
      },
    },
  },

  alias: {
    // 'aws-amplify': 'aws-amplify/lib/',
    '@aws-amplify/core': '@aws-amplify/core/lib',
    '@aws-amplify/auth': '@aws-amplify/auth/lib',
  }
})

```
I've slightly modified [the one in the fix branch of the @sweepy84 's repository](https://github.com/sweepy84/nuxt3/blob/fix/nuxt.config.ts) that I referred to in my comment on `polyfills.js`. Removed `meta` settings.

It's different from yours, but it's also solved.
Now it seems to work without the `vite` settings in this file. Maybe it was solved by upgrading Nuxt3 or Vite.
On the other hand, it seems that the `alias` setting is still necessary.
I haven't tried it, but other Amplify modules may work if you register alias as well.

I hope this information will be useful to you and other Nuxt3 users too.
Thank you.

**Reply by ennioVisco** [reply]
Thanks, a lot @mt-u. Your sweepy84's change of `nuxt.config.ts` is indeed more reliable (mine was still breaking some dependencies), so I updated the sandbox and the walkthrough accordingly.
About the `meta` settings, it is indeed alternative to the `<Script />` tag, so it makes sense that you removed it. I can confirm that the `vite.resolve.alias` block is not needed anymore, I guess it now is forwarded by nuxt directly.

**Reply by rvmourik** [reply]
@ennioVisco and @mt-u is it still working on your end with the latest nuxt version and amplify packages?

I can get it partially to work, the authenticator component can be used and logging in is also working. But as soon as I access the `Auth` object and provide it in the plugin or so even console.log it from the plugin it produces this error:

```js
global is not defined
    at node_modules/amazon-cognito-identity-js/node_modules/buffer/index.js
```
In a previous project with older versions this error didn't exist, so something seems changed but I can't figure it out. It looks like the polyfill we use through the `<Script>` component or nuxt.config.ts `meta` property is not loading in time for `Auth` object to be available. Same applies if I access the `useAuthenticator` composable from `amplify-js` for example in the `app.vue`

**Reply by ennioVisco** [reply]
> @ennioVisco and @mt-u is it still working on your end with the latest nuxt version and amplify packages?
> 
> I can get it partially to work, the authenticator component can be used and logging in is also working. But as soon as I access the `Auth` object and provide it in the plugin or so even console.log it from the plugin it produces this error:
> 
> ```js
> global is not defined
>     at node_modules/amazon-cognito-identity-js/node_modules/buffer/index.js
> ```
> 
> In a previous project with older versions this error didn't exist, so something seems changed but I can't figure it out. It looks like the polyfill we use through the `<Script>` component or nuxt.config.ts `meta` property is not loading in time for `Auth` object to be available. Same applies if I access the `useAuthenticator` composable from `amplify-js` for example in the `app.vue`

Quite annoyingly it doesn't, and I currently don't have a definitive solution. If you are *not* using `ssr`, the solution now simply is:
```ts
// nuxt.config.ts
export default defineConfig({
  // ...
  alias: {
    // ...
    './runtimeConfig': './runtimeConfig.browser',
  },
  vite: {
    // ...
    define: {
      "global": {},
    }
  }
});
```
However, if you are using `ssr`, the `define` flag makes the static generation fail (Amplify's sdk works out of the box with the alias on the client side).

**Reply by rvmourik** [reply]
Thanks for the quick reply, that is much appreciated!

I am not using `ssr` for this project so that is fine, is the polyfill also still necessary or is the define global enough?

**Reply by ennioVisco** [reply]
> Thanks for the quick reply, that is much appreciated!
> 
> I am not using `ssr` for this project so that is fine, is the polyfill also still necessary or is the define global enough?

I updated the post with the *FULL* flags that are needed. Please let us know if it worked (a reaction is good enough).

**Reply by ennioVisco** [reply]
I apologize, it seems the problem is related to the build stage (whether with `ssr` or not doesn't change). 
I created a related issue #4916.

**Reply by rvmourik** [reply]
@ennioVisco I found a work-around for the build issue:

**Replace**
```
vite: {
    define: {
      global: {}
    }
}
```

**With**
```
vite: {
    define: {
      'window.global': {}
    }
}
```

AWS is looking for global in the window, so I thought to try this out. This way, global doesn't interfere with global/globalThis.

Not ideal, but I can build my project again :-)

**Reply by ennioVisco** [reply]
Great finding @rvmourik, how don't know how I didn't think about that. This makes it much more usable, also in the short term! Thanks 👏

**Reply by rvmourik** [reply]
Could not have done it without your work in the first place, so glad we figured it out together!

**Reply by ennioVisco** [reply]
> Could not have done it without your work in the first place, so glad we figured it out together!

Sadly, your fix allows to complete the build but it still is not a valid build.
Hopefully one day they'll find a fix from Vite to this export issue.

**Reply by rvmourik** [reply]
> > Could not have done it without your work in the first place, so glad we figured it out together!
> 
> Sadly, your fix allows to complete the build but it still is not a valid build. Hopefully one day they'll find a fix from Vite to this export issue.

In what way do you mean "not a valid build", doesn't the application work after building it?

**Reply by ennioVisco** [reply]
> > > Could not have done it without your work in the first place, so glad we figured it out together!
> > 
> > 
> > Sadly, your fix allows to complete the build but it still is not a valid build. Hopefully one day they'll find a fix from Vite to this export issue.
> 
> In what way do you mean "not a valid build", doesn't the application work after building it?

Turns out it was an unrelated issue (reported [here](https://github.com/nuxt/framework/issues/5196))! It works perfectly :)

**Reply by rvmourik** [reply]
> > > > Could not have done it without your work in the first place, so glad we figured it out together!
> > > 
> > > 
> > > Sadly, your fix allows to complete the build but it still is not a valid build. Hopefully one day they'll find a fix from Vite to this export issue.
> > 
> > 
> > In what way do you mean "not a valid build", doesn't the application work after building it?
> 
> Turns out it was an unrelated issue (reported [here](https://github.com/nuxt/framework/issues/5196))! It works perfectly :)

Glad to here it :-)

**Comment by k-vikram** [comment]
Workaround for _vite + react + TS_ projects! Paste this in `index.html` :

```
<script>
    if (global === undefined) {
      var global = window;
    }
  </script> 
```

**Update:**

Found a better better fix for **Vite Projects**. Edit the `vite.config.[js/ts]` file with the following:

```
export default defineConfig({
  define: {
    "global": {},
  },
});
```
Works like a charm!

**Reply by BabyDino** [reply]
This gives me a `Cannot read property 'meta' of undefined` error

**Reply by RamonEspinosa** [reply]
I'm just going to leave a note here for other devs who get an **error** while **using the second solution**:

The second solution worked fine for me while I was in developing and testing. However, when building I got an error that looks like this:
```
[commonjs--resolver] Unexpected token (13:2) in /codebuild/output/src728824894/src/node_modules/antd/es/locale/default.js
file: /codebuild/output/src728824894/src/node_modules/antd/es/locale/default.js:13:2
11:   TimePicker: TimePicker,
12:   Calendar: Calendar,
13:   global: {
      ^
14:     placeholder: 'Please select'
15:   },
```
The [documentation](https://vitejs.dev/config/shared-options.html#define) for define says: 
> Define global constant replacements. Entries will be defined as globals during dev and statically replaced during build.

So basically what's happening is that vite searches and replaces whatever you specify in the `define` property. In my case, I'm using ant design and the source code for it has the `global` word somewhere in its i18n configuration, which then vite replaces for `{}`, thus causing an error.

**Reply by jsancho** [reply]
thanks for the heads up @RamonEspinosa 
because using this method
```
export default defineConfig({
  define: {
    "global": {},
  },
});
```
breaks for me, as this makes the `vite build` process incompatible with Fluent UI v9 `@fluentui/react-components`
```
Could not resolve "./{}/index" from "node_modules/@fluentui/tokens/lib/index.js"
file: C:/source/lynx-web-vite/node_modules/@fluentui/tokens/lib/index.js
error during build:
RollupError: Could not resolve "./{}/index" from "node_modules/@fluentui/tokens/lib/index.js"
    at error (file:///C:/source/lynx-web-vite/node_modules/vite/node_modules/rollup/dist/es/shared/rollup.js:1967:30)
    at ModuleLoader.handleInvalidResolvedId (file:///C:/source/lynx-web-vite/node_modules/vite/node_modules/rollup/dist/es/shared/rollup.js:23095:24)
    at file:///C:/source/lynx-web-vite/node_modules/vite/node_modules/rollup/dist/es/shared/rollup.js:23058:26
error Command failed with exit code 1.
```
Method 1, modifying index.html, works just fine :)

**Reply by LiamPClancy** [reply]
thank you @k-vikram the second option was working for me until I decided to use Charka UI - 
the first option is working as a solution at this stage.

**Comment by Comelpho** [comment]
@k-vikram 
Thank you so much!
I connected AWS-SDK s3 to nuxt3 and there was an error global is not defined. I added to nuxt.config.ts
```
vite:{
      define:{
        global:{}
      }
    },
```
and everything worked

**Reply by Maikpwwq** [reply]
This works for me. I have no more in my Nuxt project `ReferenceError: global is not defined.`
